nuxeo-services/nuxeo-platform-directory/nuxeo-platform-directory-test/src/test/java/org/nuxeo/directory/test/DirectoryFeature.java

```diff
@@ -108,23 +110,30 @@ public void beforeSetup(FeaturesRunner runner) throws Exception {
         allDirectoryData = new HashMap<>();
         DirectoryService directoryService = Framework.getService(DirectoryService.class);
 
-        for (Directory dir : directoryService.getDirectories()) {
-            // Do not save multi-directories as subdirectories will be saved
-            if (dir.isReadOnly() || dir instanceof MultiDirectory) {
-                continue;
+        loginStack = ClientLoginModule.getThreadLocalLogin();
+        loginStack.push(new SystemPrincipal(null), null, null);
+
+        try {
+            for (Directory dir : directoryService.getDirectories()) {
+                // Do not save multi-directories as subdirectories will be saved
+                if (dir.isReadOnly() || dir instanceof MultiDirectory) {
+                    continue;
+                }
+                try (Session session = dir.getSession()) {
+                    session.setReadAllColumns(true); // needs to fetch the password too
+                    List<DocumentModel> entries = session.query(Collections.emptyMap(), Collections.emptySet(),
+                            Collections.emptyMap(), true); // fetch references
+                    Map<String, Map<String, Object>> data = entries.stream().collect(
+                            Collectors.toMap(DocumentModel::getId, entry -> entry.getProperties(dir.getSchema())));
+                    allDirectoryData.put(dir.getName(), data);
+                }
             }
-            try (Session session = dir.getSession()) {
-                session.setReadAllColumns(true); // needs to fetch the password too
-                List<DocumentModel> entries = session.query(Collections.emptyMap(), Collections.emptySet(),
-                        Collections.emptyMap(), true); // fetch references
-                Map<String, Map<String, Object>> data = entries.stream().collect(
-                        Collectors.toMap(DocumentModel::getId, entry -> entry.getProperties(dir.getSchema())));
-                allDirectoryData.put(dir.getName(), data);
+            if (TransactionHelper.isTransactionActiveOrMarkedRollback()) {
+                TransactionHelper.commitOrRollbackTransaction();
+                TransactionHelper.startTransaction();
             }
-        }
-        if (TransactionHelper.isTransactionActiveOrMarkedRollback()) {
-            TransactionHelper.commitOrRollbackTransaction();
-            TransactionHelper.startTransaction();
+        } catch (Exception e) {
+            loginStack.pop();
```

Comment:
Shouldn't this be a `finally` instead of a `catch`?

Reply:
The idea here is to log out if an exception occurs while initializing the directories. If everything went ok, the log out will done after the test has been done in the _afterTeardown_ method.

Reply:
Then you need to rethrow the exception, don't you?

Reply:
Yes you're right, I'm adding it.